src/blight/actions/find_inputs.py

```diff
@@ -0,0 +1,104 @@
+"""
+The `FindInputs` action.
+"""
+
+from pathlib import Path
+from typing import List, Optional
+
+from pydantic import BaseModel
+
+from blight.action import Action
+from blight.constants import INPUT_SUFFIX_KIND_MAP
+from blight.enums import InputKind
+from blight.tool import Tool
+from blight.util import flock_append
+
+
+class Input(BaseModel):
+    """
+    Represents a single input to a tool.
+    """
+
+    kind: InputKind
+    """
+    The kind of input.
+    """
+
+    prenormalized_path: str
+    """
+    The path to the input, as passed directly to the tool itself.
+
+    This copy of the path may be relative; consumers should prefer
+    `path` for an absolute copy.
+    """
+
+    path: Path
+    """
+    The path to the input, as created by the tool.
+
+    **NOTE**: This path may not actually exist, as a build system may arbitrarily
+    choose to rename or relocate any inputs consumed by individual tools.
+    """
+
+    store_path: Optional[Path]
+    """
+    An optional stable path to the input, as preserved by the `FindInputs` action.
+
+    `store_path` is only present if the `store=/some/dir/` setting is passed in the
+    `FindInputs` configuration **and** the tool actually produces the expected input.
+    """
+
+    content_hash: Optional[str]
+    """
+    A SHA256 hash of the input's content.
+
+    `content_hash` is only present if the `store=/some/dir/` setting is passed in the
+    `FindInputs` configuration **and** the tool actually consumes the expected input.
+    """
+
+
+class InputsRecord(BaseModel):
+    """
+    Represents a single `FindInputs` record. Each record contains a representation
+    of the tool invocation that consumes the inputs, as well as the list of `Input`s
+    associated with the invocation.
+    """
+
+    tool: Tool
+    """
+    The `Tool` invocation.
+    """
+
+    inputs: List[Input]
+    """
+    A list of `Input`s.
+    """
+
+    class Config:
+        arbitrary_types_allowed = True
+        json_encoders = {Tool: lambda t: t.asdict()}
+
+
+class FindInputs(Action):
+    def before_run(self, tool: Tool) -> None:
+        inputs = []
+        for input in tool.inputs:
+            input_path = Path(input)
+            if not input_path.is_absolute():
+                input_path = tool.cwd / input_path
+
+            kind = INPUT_SUFFIX_KIND_MAP.get(input_path.suffix, InputKind.Unknown)
```

Comment:
Should probably do `input_path.suffix.lower()` to handle things like `.C`, etc.

Reply:
Yeah, let's do a followup for that. I bet `FindOutputs` is also probably wrong.

Reply:
I have some similar code to this in PASTA. Here's what I do:

```c++
    std::vector<std::string> new_argv;
    for (const char *arg : job_args) {

      // Try to fixup any remaining paths.
      llvm::StringRef a(arg);

      // Try to look for things that look file file names, then see if they are
      // file names, and if so, make them absolute paths.
      if (a.contains("./") || a.endswith_insensitive(".o") ||
          a.endswith_insensitive(".cc") || a.endswith_insensitive(".hh") ||
          a.endswith_insensitive(".cpp") || a.endswith_insensitive(".hpp") ||
          a.endswith_insensitive(".c++") || a.endswith_insensitive(".h++") ||
          a.endswith_insensitive(".cxx") || a.endswith_insensitive(".hxx") ||
          a.endswith_insensitive(".c") || a.endswith_insensitive(".h") ||
          a.endswith_insensitive(".gcno") || a.endswith_insensitive(".pch") ||
          a.endswith_insensitive(".s") || a.endswith_insensitive(".asm") ||
          a.endswith_insensitive(".mm") || a.endswith_insensitive(".d") ||
          a.endswith_insensitive(".sdk")) {

        if (auto maybe_info = fs.Stat(arg); maybe_info.Succeeded()) {
          new_argv.emplace_back(
              maybe_info.TakeValue().real_path.generic_string());
          continue;
        }
      }

      new_argv.emplace_back(arg);
    }
```

Reply:
`-include foo.h`.

Reply:
😭